src/java/org/apache/cassandra/cache/ChunkCache.java

```diff
@@ -158,88 +195,262 @@ public boolean equals(Object obj)
-            return buffer.asIntBuffer();
+            super(offset);
+            this.pages = new long[buffers.length];
+            this.attachments = new Object[buffers.length];
+            for (int i = 0; i < buffers.length; ++i)
+            {
+                pages[i] = MemoryUtil.getAddress(buffers[i]);
+                attachments[i] = MemoryUtil.getAttachment(buffers[i]);
+            }
         }
 
-        @Override
-        public LongBuffer longBuffer()
+        void releaseBuffers()
+        {
+            for (int i = 0; i < pages.length; ++i)
+                bufferPool.put(MemoryUtil.allocateByteBuffer(pages[i], PageAware.PAGE_SIZE, PageAware.PAGE_SIZE, ByteOrder.BIG_ENDIAN, attachments[i]));
+        }
+
+        void read(ChunkReader file)
         {
-            assert references.get() > 0;
-            // this does an implicit duplicate(), so we need to expose it directly to avoid doing it twice unnecessarily
-            return buffer.asLongBuffer();
+            bytesRead = readScattering(file, offset, capacity(), pages);
         }
 
+        @Nullable
+        Buffer getBuffer(long position)
+        {
+            int index = PageAware.pageNum(position - offset);
+            Preconditions.checkArgument(index >= 0 && index < pages.length, "Invalid position: %s, index: %s", position, index);
+
+            long pageAlignedPosition = PageAware.pageStart(position);
+            int bufferSize = Math.min(PageAware.PAGE_SIZE, Math.toIntExact(bytesRead - (index * PageAware.PAGE_SIZE)));
+            assert bufferSize > 0 && bufferSize <= PageAware.PAGE_SIZE : "Wrong buffer size: " + bufferSize + " at position " + position;
+
+            return new Buffer(pages[index], pageAlignedPosition, bufferSize);
+        }
+
+        public int capacity()
+        {
+            return pages.length * PageAware.PAGE_SIZE;
+        }
+
+        class Buffer implements Rebufferer.BufferHolder
+        {
+            private final long address;
+            private final long offset;
+            private final int limit;
+
+
+            public Buffer(long address, long offset, int limit)
+            {
+                this.address = address;
+                this.offset = offset;
+                this.limit = limit;
+            }
+
+            @Override
+            public ByteBuffer buffer()
+            {
+                assert isReferenced() : "Already unreferenced";
+                return MemoryUtil.allocateByteBuffer(address, limit, PageAware.PAGE_SIZE, ByteOrder.BIG_ENDIAN, null);
+            }
+
+            @Override
+            public long offset()
+            {
+                return offset;
+            }
+
+            @Override
+            public void release()
+            {
+                MultiRegionChunk.this.release();
+            }
+        }
+    }
+
+    /**
+     * A chunk with a single memory region. This can be used for reading chunks of up to PageAware.PAGE_SIZE but note
```

Comment:
The "This can be used for reading chunks of up to PageAware.PAGE_SIZE" sentence here is a bit confusing, given we use this for reading larger chunks when we're able to get contiguous regions.